distributed/deploy/cluster.py

```diff
@@ -564,3 +565,51 @@ def __eq__(self, other):
 
     def __hash__(self):
         return id(self)
+
+    async def _wait_for_workers(self, n_workers=0, timeout=None):
+        info = await self.get_worker_status()
+        self._scheduler_identity = info
+        if timeout:
+            deadline = time() + parse_timedelta(timeout)
+        else:
+            deadline = None
+
+        def running_workers(info):
+            return len(
+                [
+                    ws
+                    for ws in info["workers"].values()
+                    if ws["status"] == Status.running.name
+                ]
+            )
+
+        while n_workers and running_workers(info) < n_workers:
+            if deadline and time() > deadline:
+                raise TimeoutError(
+                    "Only %d/%d workers arrived after %s"
+                    % (running_workers(info), n_workers, timeout)
+                )
+            await asyncio.sleep(0.1)
+
+            info = await self.get_worker_status()
+            self._scheduler_identity = info
+
+    async def get_worker_status(self):
+        comm = await self.scheduler_comm.live_comm()
+        comm.name = "Cluster worker status"
+        await comm.write({"op": "identity"})
+        self.scheduler_info = await comm.read()
+        return SchedulerInfo(self.scheduler_info)
```

Comment:
This feels a little muddy as it is updating the `scheduler_info` which is then used to read the worker status outside of the function. There is also a bunch of duplication here from the `_start` method. Perhaps this should be renamed to `update_scheduler_info` or similar and used in both places.

Also given that it opens a comm and uses it briefly perhaps it would be better to use a context manager?

Reply:
Yes, it's much better after a little refactoring.

I tried to use a context manager, but I don't think it works here as the cluster needs to keep scheduler_comm open or we start getting errors in other places, e.g. I was getting a RuntimeError during startup when it called get_metadata. Or did I misunderstand?